### PR TITLE
reboot, snapshot, retsore action and handler redirect

### PR DESCRIPTION
**Pull request by yb01**
reboot and snapshot action from the openstack API redirects.

local test:
```
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# 
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# kubectl get pods -AT
TENANT   NAMESPACE     NAME                              HASHKEY               READY   STATUS    RESTARTS   AGE
system   kube-system   coredns-default-649bd7994-n7zxx   2554247618566393352   1/1     Running   0          17s
system   kube-system   kube-dns-554c5866fc-rrrtt         7966972732709566351   3/3     Running   0          17s
system   kube-system   testvm                            9075866572257574198   1/1     Running   0          5s
system   kube-system   virtlet-m9s8t                     1865010651528817676   3/3     Running   0          17s
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# curl -L -X POST http://127.0.0.1:8080/servers/testvm/action -H "Content-Type: application/json" -H "openstack: true" -d @./createImage.json -v | jq
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1...
* TCP_NODELAY set
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> POST /servers/testvm/action HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/7.58.0
> Accept: */*
> Content-Type: application/json
> openstack: true
> Content-Length: 115
> 
} [115 bytes data]
* upload completely sent off: 115 out of 115 bytes
< HTTP/1.1 307 Temporary Redirect
< Cache-Control: no-cache, private
< Location: /api/v1/tenants/system/namespaces/kube-system/pods/testvm/action
< Date: Wed, 05 Jan 2022 07:27:26 GMT
< Content-Length: 0
< 
100   115    0     0  100   115      0   112k --:--:-- --:--:-- --:--:--  112k
* Connection #0 to host 127.0.0.1 left intact
* Issue another request to this URL: 'http://127.0.0.1:8080/api/v1/tenants/system/namespaces/kube-system/pods/testvm/action'
* Found bundle for host 127.0.0.1: 0x558cbc076ae0 [can pipeline]
* Re-using existing connection! (#0) with host 127.0.0.1
* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> POST /api/v1/tenants/system/namespaces/kube-system/pods/testvm/action HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/7.58.0
> Accept: */*
> Content-Type: application/json
> openstack: true
> Content-Length: 115
> 
} [115 bytes data]
* upload completely sent off: 115 out of 115 bytes
< HTTP/1.1 201 Created
< Cache-Control: no-cache, private
< Content-Type: application/json
< Date: Wed, 05 Jan 2022 07:27:26 GMT
< Content-Length: 38
< 
{ [38 bytes data]
100   153  100    38  100   115   5428  16428 --:--:-- --:--:-- --:--:-- 21856
* Connection #0 to host 127.0.0.1 left intact
{
  "ImageId": "snapshot-1641367646"
}
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# curl -L -X POST http://127.0.0.1:8080/servers/testvm/action -H "Content-Type: application/json" -H "openstack: true" -d @./rebuild.json -v | jq
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1...
* TCP_NODELAY set
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> POST /servers/testvm/action HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/7.58.0
> Accept: */*
> Content-Type: application/json
> openstack: true
> Content-Length: 118
> 
} [118 bytes data]
* upload completely sent off: 118 out of 118 bytes
< HTTP/1.1 307 Temporary Redirect
< Cache-Control: no-cache, private
< Location: /api/v1/tenants/system/namespaces/kube-system/pods/testvm/action
< Date: Wed, 05 Jan 2022 07:27:31 GMT
< Content-Length: 0
< 
100   118    0     0  100   118      0   115k --:--:-- --:--:-- --:--:--  115k
* Connection #0 to host 127.0.0.1 left intact
* Issue another request to this URL: 'http://127.0.0.1:8080/api/v1/tenants/system/namespaces/kube-system/pods/testvm/action'
* Found bundle for host 127.0.0.1: 0x55b819699ae0 [can pipeline]
* Re-using existing connection! (#0) with host 127.0.0.1
* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> POST /api/v1/tenants/system/namespaces/kube-system/pods/testvm/action HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/7.58.0
> Accept: */*
> Content-Type: application/json
> openstack: true
> Content-Length: 118
> 
} [118 bytes data]
* upload completely sent off: 118 out of 118 bytes
< HTTP/1.1 201 Created
< Cache-Control: no-cache, private
< Content-Type: application/json
< Date: Wed, 05 Jan 2022 07:27:31 GMT
< Content-Length: 74
< 
{ [74 bytes data]
100   192  100    74  100   118  14800  23600 --:--:-- --:--:-- --:--:-- 38400
* Connection #0 to host 127.0.0.1 left intact
{
  "ServerId": "",
  "ImageId": "restore-1641367651",
  "CreatedAt": ""
}
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# 
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# kubectl get actions -AT
TENANT   NAMESPACE     NAME                  PODNAME   COMPLETE   ERROR
system   kube-system   restore-1641367651    testvm    Yes        
system   kube-system   snapshot-1641367646   testvm    Yes        
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# 
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 1]
```diff
@@ -101,19 +101,27 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 		decoder := scope.Serializer.DecoderToVersion(s.Serializer, scope.HubGroupVersion)
 
 		if openstack.IsOpenstackRequest(req) {
-			klog.V(4).Infof("convert the request body to arktos vm pod")
+			klog.V(4).Infof("convert the request body to Arktos request")
```
Comment (Sindica) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:104]: Nit: there is no need for Infof

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go [hunk 2]
```diff
@@ -140,16 +140,24 @@ func transformResponseObject(ctx context.Context, scope *RequestScope, trace *ut
 
 	// transform to Openstack response structure
 	if openstack.IsOpenstackRequest(req) {
-		obj = convertObjectToOpenstackResponse(obj)
+		klog.V(6).Infof("Transform to openstack response object. object: %s", obj)
```
Comment (Sindica) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go:143]: %v or %#v?

File: staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go [hunk 3]
```diff
@@ -107,16 +206,78 @@ func ConvertToOpenstackRequest(body []byte) ([]byte, error) {
...
+		s := OpenstackCreateImageResponse{ImageId: o.Details.Name}
+		return &s
+	}
+
+	if strings.Contains(o.Message, RESTORE) {
+		s := OpenstackRebuildResponse{ImageId: o.Details.Name, CreatedAt: time.Now().String()}
+		return &s
+	}
+
+	return o
```
Comment (Sindica) [inline staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go:278]: Perhaps log to indicate that this action does not support for now?

**Review by Sindica — APPROVED** [review]
lgtm. Minor comment below.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go (hunk 1 above)
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: debug log. removed for now.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go (hunk 2 above)
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go]: fixed. thanks for the catch.

File: staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go (hunk 3 above)
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go:278]: the log is in the ConvertActionFromOpenstackRequest() function as the default of the action operations.

**Pull request merged**